[CI] org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingServiceTests.testReferenceCountingForPipeline

<!--
Please fill out the following information, and ensure you have attempted
to reproduce locally
-->

**Build scan**: https://gradle-enterprise.elastic.co/s/32ejz4ocpihbo

**Repro line**:
`./gradlew ':x-pack:plugin:ml:test' --tests "org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingServiceTests.testReferenceCountingForPipeline" -Dtests.seed=4887241A54A45DE6 -Dtests.security.manager=true -Dtests.locale=he-IL -Dtests.timezone=Asia/Almaty -Druntime.java=11`

**Reproduces locally?**:
No.

**Applicable branches**:
`master`

**Failure history**:
N/A

**Failure excerpt**:
```
java.lang.AssertionError: expected:<2> but was:<3>
...
at __randomizedtesting.SeedInfo.seed([4887241A54A45DE6:8478EDE36A323A78]:0)
...
at org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingServiceTests.testReferenceCountingForPipeline(ModelLoadingServiceTests.java:507)
```
https://gradle-enterprise.elastic.co/s/32ejz4ocpihbo/tests/:x-pack:plugin:ml:test/org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingServiceTests/testReferenceCountingForPipeline#1

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingServiceTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingServiceTests.java
@@ -452,7 +452,7 @@ public class ModelLoadingServiceTests extends ESTestCase {
         });
     }
 
-    public void testReferenceCounting() throws ExecutionException, InterruptedException, IOException {
+    public void testReferenceCounting() throws Exception {
         String modelId = "test-reference-counting";
         withTrainedModel(modelId, 1L);
 
@@ -469,24 +469,24 @@ public class ModelLoadingServiceTests extends ESTestCase {
 
         PlainActionFuture<LocalModel> forPipeline = new PlainActionFuture<>();
         modelLoadingService.getModelForPipeline(modelId, forPipeline);
-        LocalModel model = forPipeline.get();
-        assertEquals(2, model.getReferenceCount());
+        final LocalModel model = forPipeline.get();
+        assertBusy(() -> assertEquals(2, model.getReferenceCount()));
 
         PlainActionFuture<LocalModel> forSearch = new PlainActionFuture<>();
         modelLoadingService.getModelForPipeline(modelId, forSearch);
-        model = forSearch.get();
-        assertEquals(3, model.getReferenceCount());
+        forSearch.get();
+        assertBusy(() -> assertEquals(3, model.getReferenceCount()));
 
         model.release();
-        assertEquals(2, model.getReferenceCount());
+        assertBusy(() -> assertEquals(2, model.getReferenceCount()));
 
         PlainActionFuture<LocalModel> forSearch2 = new PlainActionFuture<>();
         modelLoadingService.getModelForPipeline(modelId, forSearch2);
-        model = forSearch2.get();
-        assertEquals(3, model.getReferenceCount());
+        forSearch2.get();
+        assertBusy(() -> assertEquals(3, model.getReferenceCount()));
     }
 
-    public void testReferenceCountingForPipeline() throws ExecutionException, InterruptedException, IOException {
+    public void testReferenceCountingForPipeline() throws Exception {
         String modelId = "test-reference-counting-for-pipeline";
         withTrainedModel(modelId, 1L);
 
@@ -503,18 +503,17 @@ public class ModelLoadingServiceTests extends ESTestCase {
 
         PlainActionFuture<LocalModel> forPipeline = new PlainActionFuture<>();
         modelLoadingService.getModelForPipeline(modelId, forPipeline);
-        LocalModel model = forPipeline.get();
-        assertEquals(2, model.getReferenceCount());
+        final LocalModel model = forPipeline.get();
+        assertBusy(() -> assertEquals(2, model.getReferenceCount()));
 
         PlainActionFuture<LocalModel> forPipeline2 = new PlainActionFuture<>();
         modelLoadingService.getModelForPipeline(modelId, forPipeline2);
-        model = forPipeline2.get();
-        assertEquals(3, model.getReferenceCount());
+        forPipeline2.get();
+        assertBusy(() -> assertEquals(3, model.getReferenceCount()));
 
         // will cause the model to be evicted
         modelLoadingService.clusterChanged(ingestChangedEvent());
-
-        assertEquals(2, model.getReferenceCount());
+        assertBusy(() -> assertEquals(2, model.getReferenceCount()));
     }
 
     public void testReferenceCounting_ModelIsNotCached() throws ExecutionException, InterruptedException {

